File: backend/app/mapper/medication.py
import hashlib
import re
from collections import OrderedDict
from datetime import date, timedelta

from app.fhir import systems
from app.mapper.condition import _normalize_icd10_cm


def _med_status(authored_iso: str, duration_days) -> str:
    """Best-effort active vs completed decision for a MedicationRequest.

    NHI 健保存摺 gives us authored date (ROC date converted upstream to ISO)
    and `order_drug_day` — total days supplied. The medication is
    "active" while  authored_date + duration > today  ; otherwise the
    prescription's supply has run out and we mark it "completed".

    When duration is missing or unparseable we treat anything older than
    90 days as completed — typical refill cadence — and anything more
    recent as active, since most patients are still on the drug.
    Conservative defaults so the SMART app's "current medications" view
    doesn't surface year-old prescriptions as if they're current.
    """
    if not authored_iso:
        return "completed"
    try:
        d = date.fromisoformat(authored_iso[:10])
    except (ValueError, TypeError):
        return "completed"
    try:
        days = int(duration_days) if duration_days not in (None, "") else None
    except (ValueError, TypeError):
        days = None
    if days is None:
        days = 90  # heuristic supply window when NHI didn't tell us
    today = date.today()
    end = d + timedelta(days=days)
    return "active" if end >= today else "completed"


def map_medications_dedup(raw_items: list[dict], patient_id: str) -> list[dict]:
    """Group-aware medication mapper that dedupes 中英 雙語 duplicates.

    NHI 健康存摺's 用藥紀錄 lists the SAME prescription multiple times because
    each drug is reported in 3 formats by different data sources:
      - "TIMOPTOL XE 0.5% OPHTHALMIC SOLUTION"           (English only)
      - "TIMOPTOL XE 0.5% OPHTHALMIC SOLUTION (青眼露…)" (Eng + 中)
      - "青眼露… (TIMOPTOL XE 0.5% OPHTHALMIC SOLUTION)" (中 + Eng)
    The current backend creates one MedicationRequest per row → 3x duplicates
    per drug per visit.

    Strategy:
      1. Compute a canonical key per drug name (longest English chunk after
         stripping parens and casefolding). All three formats above collapse
         to the same key.
      2. Group by (date, canonical_key). For each group, keep ONE entry,
         preferring the name with the most CJK characters (Chinese display
         is what clinicians read).
      3. Map each kept entry through map_medication_request.

    Falls back gracefully — drugs with no English component (rare for NHI)
    use their full name as the key, so different drugs don't collide.
    """
    by_key: OrderedDict[tuple, dict] = OrderedDict()
    for item in raw_items:
        if not isinstance(item, dict):
            continue
        drug_name = (item.get("drug_name") or "").strip()
        if not drug_name:
            continue
        date = (item.get("date") or "")[:10]
        key = (date, _canonical_drug_key(drug_name))
        existing = by_key.get(key)
        if existing is None:
            by_key[key] = item
        else:
            # Prefer the English form (fewer CJK characters). Clinicians
            # read prescriptions in English brand names (TIMOPTOL, XYZAL,
            # CURAM) — the Chinese 商品名 appended in parens is verbose
            # and adds little when scanning a med list.
            if _cjk_chars(drug_name) < _cjk_chars(existing.get("drug_name", "")):
                by_key[key] = item

    out: list[dict] = []
    for item in by_key.values():
        m = map_medication_request(item, patient_id)
        if m is not None:
            out.append(m)
    return out


def _cjk_chars(s: str) -> int:
    return sum(1 for c in (s or "") if "一" <= c <= "鿿")


# Match a "long" English chunk (≥4 chars of A-Z/0-9/punctuation common to
# drug names). Avoid matching short tokens like "D" or "PO" that appear
# inside Chinese names.
_EN_CHUNK = re.compile(r"[A-Z][A-Z0-9.%/\-\"'\s]{3,}")


def _canonical_drug_key(name: str) -> str:
    """Reduce a drug-name string to a stable canonical key.

    Strategy: extract the longest English fragment, then truncate at
    common separators (" - ", " / ") so a name with extra trailing
    modifiers still collapses to the brand+strength prefix.

    Examples (all map to the same key):
      "TIMOPTOL XE 0.5% OPHTHALMIC SOLUTION"
      "TIMOPTOL XE 0.5% OPHTHALMIC SOLUTION (青眼露…)"
      "青眼露… (TIMOPTOL XE 0.5% OPHTHALMIC SOLUTION)"
      → "timoptol xe 0.5% ophthalmic solution"

      "FLUCASON OPHTHALMIC SUSPENSION 0.02% (FLUOROMETHOLONE)"
      "護康視懸濁點眼液 0.02% (FLUCASON OPHTHALMIC SUSPENSION 0.02% - FLUOROMETHOLONE)"
      → "flucason ophthalmic suspension 0.02%"
    """
    s = (name or "").upper()
    chunks = _EN_CHUNK.findall(s)
    if not chunks:
        return (name or "").strip().lower()
    longest = max(chunks, key=len).strip()
    # Drop trailing dashed/slashed modifiers (typically the generic name
    # appended after the brand+strength).
    for sep in (" - ", " – ", " / "):
        if sep in longest:
            longest = longest.split(sep)[0]
    return re.sub(r"\s+", " ", longest).lower().strip()


def map_medication_request(raw: dict, patient_id: str) -> dict | None:
    """Convert scraped prescription dict → FHIR R4 MedicationRequest resource.

    Returns None when the raw item has no drug name (e.g. a ghost entry
    extracted from a visit header or a non-drug order row) so the caller
    can filter it out.
    """
    drug_name = (raw.get("drug_name") or "").strip()
    if not drug_name:
        return None  # skip visit-header / non-drug-order ghost entries

    # Use canonical key (not raw drug_name) for the stable id so the three
    # NHI 中英 variants of the same drug collapse to one FHIR resource even
    # when they arrive in separate /sync/upload-html POSTs.
    med_id = _stable_id(patient_id, _canonical_drug_key(drug_name), raw.get("date", ""))

    # Prefer the actual prescription code (e.g. NHI 健保藥品代碼 "BC23792100")
    # over reusing the display name as a code. The HIS-specific OID lets
    # downstream tools recognise the namespace.
    drug_code = (raw.get("code") or "").strip()
    coding = {
        "system": systems.NHI_DRUG_CODE if drug_code else systems.HIS_LOCAL_MEDICATION_CODE,
        "code": drug_code or drug_name,
        "display": drug_name,
    }

    resource: dict = {
        "resourceType": "MedicationRequest",
        "id": med_id,
        "meta": {"versionId": "1", "source": "nhi-fhir-bridge/scraper"},
        "status": _med_status(raw.get("date", ""), raw.get("duration_days")),
        "intent": "order",
        "medicationCodeableConcept": {
            "coding": [coding],
            "text": drug_name,
        },
        "subject": {"reference": f"Patient/{patient_id}"},
    }

    if raw.get("date"):
        resource["authoredOn"] = raw["date"] + "T00:00:00+08:00"

    # ATC-style drug class (NHI's `act` field — "ANTIHISTAMINE DRUGS" etc.).
    # FHIR puts pharmacological class on MedicationRequest.category.
    drug_class = (raw.get("drug_class") or "").strip()
    if drug_class:
        resource["category"] = [{"text": drug_class}]

    # Hospital/院所 → MedicationRequest.requester.display. NHI's data is
    # billing-tracked: the hospital that submitted the medication claim
    # is the prescriber+dispenser (for outpatient meds in healthcare-bank
    # context the two are the same). Setting this is also how the
    # post-mapping linker finds the matching Encounter by (hospital, date).
    hospital = (raw.get("hospital") or "").strip()
    if hospital:
        resource["requester"] = {"display": hospital}

    # Dosage instruction — only emit when the source actually gave us
    # dose / frequency / route. NHI's medication-list endpoint provides
    # none of these (only qty + duration on dispenseRequest), so for
    # NHI-sourced data this block stays empty. Other HIS adapters that
    # populate these fields still get a structured dosage out.
    dosage: dict = {}
    parts = [str(raw[k]) for k in ("dose", "unit", "frequency") if raw.get(k)]
    if parts:
        dosage["text"] = " ".join(parts)
    if raw.get("route"):
        dosage["route"] = {
            "coding": [{"system": "http://snomed.info/sct", "display": raw["route"]}]
        }
    if dosage:
        resource["dosageInstruction"] = [dosage]

    # dispenseRequest now carries BOTH the total quantity and the supply
    # duration when NHI gives us both — previously we only kept duration.
    # No unit on quantity: NHI doesn't tell us whether 14 is tablets,
    # capsules, mL, etc. — adding "{tablet}" would be a fabrication.
    dr: dict = {}
    qty_raw = raw.get("quantity")
    if qty_raw not in (None, ""):
        try:
            dr["quantity"] = {"value": float(str(qty_raw).replace(",", ""))}
        except (ValueError, TypeError):
            pass
    if raw.get("duration_days"):
        try:
            days = int(raw["duration_days"])
            dr["expectedSupplyDuration"] = {
                "value": days,
                "unit": "days",
                "system": "http://unitsofmeasure.org",
                "code": "d",
            }
        except (ValueError, TypeError):
            pass
    if dr:
        resource["dispenseRequest"] = dr

    # Indication: prefer the structured ICD code when present, fall back
    # to the free-text name. Both arrive from NHI as separate fields:
    # icd9cm_CODE → indication_code, icd9cm_CODE_CNAME → indication.
    indication = (raw.get("indication") or "").strip()
    indication_code = (raw.get("indication_code") or "").strip()
    if indication or indication_code:
        rc: dict = {}
        if indication_code:
            rc["coding"] = [
                {
                    # NHI's indication codes are ICD-10-CM (e.g. E11.22), not
                    # the base ICD-10 ValueSet — switching the system URL
                    # so the validator's terminology lookup resolves correctly.
                    # Also normalise NHI's no-dot form ('E1122' → 'E11.22')
                    # to match the canonical ValueSet.
                    "system": systems.ICD_10_CM,
                    "code": _normalize_icd10_cm(indication_code),
                    "display": indication or indication_code,
                }
            ]
        if indication:
            rc["text"] = (
                f"{indication_code} {indication}".strip() if indication_code else indication
            )
        resource["reasonCode"] = [rc]

    return resource


def _stable_id(patient_id: str, *parts: str) -> str:
    key = "|".join([patient_id, *parts])
    return hashlib.sha1(key.encode()).hexdigest()[:32]

File: packages/mapper/src/medication.ts
/**
 * MedicationRequest mapper + bilingual deduplication.
 *
 * Port of `backend/app/mapper/medication.py`. NHI 健康存摺 reports the
 * SAME prescription multiple times (English-only / Eng+中 / 中+Eng).
 * `mapMedicationsDedup` collapses these to one MedicationRequest per
 * (date, canonical-drug-key), preferring the form with more CJK chars
 * (clinicians read 商品名 first).
 */

import * as systems from "./systems";
import { normalizeIcd10Cm } from "./condition";
import { stableId } from "./helpers";

function isCjk(ch: string): boolean {
  // 一 (U+4E00) to 鿿 (U+9FFF) covers CJK Unified Ideographs.
  const cp = ch.codePointAt(0) ?? 0;
  return cp >= 0x4e00 && cp <= 0x9fff;
}

function cjkChars(s: string | null | undefined): number {
  if (!s) return 0;
  let n = 0;
  for (const ch of s) if (isCjk(ch)) n++;
  return n;
}

/**
 * Match a "long" English chunk (≥4 chars of A-Z/0-9/punctuation common
 * to drug names). Avoid matching short tokens like "D" or "PO" that
 * appear inside Chinese names.
 */
const EN_CHUNK_G = /[A-Z][A-Z0-9.%/\-"'\s]{3,}/g;

/**
 * Reduce a drug-name string to a stable canonical key. Extract the
 * longest English fragment, then truncate at common separators so a
 * name with extra trailing modifiers still collapses to brand+strength.
 *
 * Examples (all map to "timoptol xe 0.5% ophthalmic solution"):
 *   "TIMOPTOL XE 0.5% OPHTHALMIC SOLUTION"
 *   "TIMOPTOL XE 0.5% OPHTHALMIC SOLUTION (青眼露…)"
 *   "青眼露… (TIMOPTOL XE 0.5% OPHTHALMIC SOLUTION)"
 */
export function canonicalDrugKey(name: string | null | undefined): string {
  const s = (name ?? "").toUpperCase();
  const chunks = [...s.matchAll(EN_CHUNK_G)].map((m) => m[0]);
  if (chunks.length === 0) {
    return (name ?? "").trim().toLowerCase();
  }
  let longest = chunks.reduce((a, b) => (b.length > a.length ? b : a)).trim();
  for (const sep of [" - ", " – ", " / "]) {
    if (longest.includes(sep)) {
      longest = longest.split(sep)[0]!;
    }
  }
  return longest.replace(/\s+/g, " ").trim().toLowerCase();
}

/**
 * Best-effort active vs completed decision for a MedicationRequest.
 * Active while (authored_date + duration > today); otherwise completed.
 * Missing duration → assume 90-day refill window (NHI's typical cadence).
 */
export function medStatus(
  authoredIso: string | null | undefined,
  durationDays: any,
): "active" | "completed" {
  if (!authoredIso) return "completed";
  const datePart = String(authoredIso).slice(0, 10);
  const parsed = new Date(`${datePart}T00:00:00Z`);
  if (Number.isNaN(parsed.getTime())) return "completed";

  let days: number | null;
  if (durationDays === null || durationDays === undefined || durationDays === "") {
    days = null;
  } else {
    const n = Number.parseInt(String(durationDays), 10);
    days = Number.isFinite(n) ? n : null;
  }
  if (days === null) days = 90;

  const end = new Date(parsed.getTime());
  end.setUTCDate(end.getUTCDate() + days);
  // Compare date-only (today in UTC since we authoredIso is date-only).
  const today = new Date();
  today.setUTCHours(0, 0, 0, 0);
  return end >= today ? "active" : "completed";
}

/**
 * Convert one scraped prescription dict → FHIR R4 MedicationRequest.
 * Returns null when raw has no `drug_name` (caller filters out).
 */
export function mapMedicationRequest(
  raw: Record<string, any>,
  patientId: string,
): Record<string, any> | null {
  const drugName = ((raw.drug_name ?? "") as string).trim();
  if (!drugName) return null;

  // Canonical key (not raw drug_name) for stable id so the three NHI
  // 中英 variants of the same drug collapse to one FHIR resource.
  const medId = stableId(patientId, canonicalDrugKey(drugName), raw.date ?? "");

  const drugCode = ((raw.code ?? "") as string).trim();
  const coding: Record<string, string> = {
    system: drugCode ? systems.NHI_DRUG_CODE : systems.HIS_LOCAL_MEDICATION_CODE,
    code: drugCode || drugName,
    display: drugName,
  };

  // v0.8.0 bilingual: prefer 繁中 in CodeableConcept.text (patient-facing
  // display) and keep English in coding[0].display (clinical canonical).
  // Falls back to English when NHI didn't ship a Chinese name for the drug.
  const drugNameZh = ((raw.drug_name_zh ?? "") as string).trim() || drugName;

  const resource: Record<string, any> = {
    resourceType: "MedicationRequest",
    id: medId,
    meta: { versionId: "1", source: "nhi-fhir-bridge/scraper" },
    status: medStatus(raw.date ?? "", raw.duration_days),
    intent: "order",
    medicationCodeableConcept: {
      coding: [coding],
      text: drugNameZh,
    },
    subject: { reference: `Patient/${patientId}` },
  };

  if (raw.date) {
    resource.authoredOn = `${raw.date}T00:00:00+08:00`;
  }

  // Chronic prescriptions (from NHI's IHKE3307S01 慢性處方箋 list) get
  // the standard FHIR continuous-therapy marker. SMART apps recognise
  // this code and can surface "long-term medication" badges or filter
  // problem-list views. Acute prescriptions leave the field unset.
  const courseOfTherapy = ((raw.course_of_therapy ?? "") as string).trim();
  if (courseOfTherapy === "continuous") {
    resource.courseOfTherapyType = {
      coding: [
        {
          system:
            "http://terminology.hl7.org/CodeSystem/medicationrequest-course-of-therapy",
          code: "continuous",
          display: "Continuous long term therapy",
        },
      ],
      text: "Continuous long term therapy",
    };
  }

  const drugClass = ((raw.drug_class ?? "") as string).trim();
  const drugClassZh = ((raw.drug_class_zh ?? "") as string).trim();
  if (drugClass || drugClassZh) {
    const cat: Record<string, any> = {};
    if (drugClass) cat.coding = [{ display: drugClass }];
    // Patient-facing: prefer 繁中 in .text, fall back to English.
    cat.text = drugClassZh || drugClass;
    resource.category = [cat];
  }

  const hospital = ((raw.hospital ?? "") as string).trim();
  if (hospital) {
    resource.requester = { display: hospital };
  }

  // Dosage — three-level fallback:
  //   1. structured dose/unit/frequency from adapter (preferred)
  //   2. raw NHI 用法 text via dosage_text (Part 3 C6 scaffold)
  //   3. derived "${qty}/${days} days" from quantity + duration_days
  //      (Part 6 U4 — NHI 健保存摺 IHKE3306S02 doesn't expose 用法 as a
  //      separate field but DOES expose 給藥日數 + 給藥總量. The ratio
  //      is the only daily-frequency signal we get from NHI raw, so
  //      emit it as a transport-only text so clinicians see SOMETHING
  //      rather than nothing. Format intentionally generic — we don't
  //      know units (tablet vs ml) or timing (QD vs BID with halves).
  //      SMART app can choose to display raw or parse further.)
  const dosage: Record<string, any> = {};
  const parts: string[] = [];
  for (const k of ["dose", "unit", "frequency"] as const) {
    if (raw[k]) parts.push(String(raw[k]));
  }
  if (parts.length > 0) {
    dosage.text = parts.join(" ");
  } else if (raw.dosage_text) {
    const t = String(raw.dosage_text).trim();
    if (t) dosage.text = t;
  } else {
    // Level 3: derive from qty + days. Only fires when both > 0 to
    // avoid emitting "0/0 days" or partial nonsense.
    const qtyRaw = raw.quantity;
    const qtyNum =
      qtyRaw !== null && qtyRaw !== undefined && qtyRaw !== ""
        ? Number.parseFloat(String(qtyRaw).replace(/,/g, ""))
        : Number.NaN;
    const daysNum = Number.parseInt(String(raw.duration_days ?? 0), 10);
    if (Number.isFinite(qtyNum) && qtyNum > 0 && Number.isFinite(daysNum) && daysNum > 0) {
      // Render qty without trailing zeros (3.0 → "3", 1.5 stays).
      const qtyStr = Number.isInteger(qtyNum) ? String(qtyNum) : String(qtyNum);
      const perDay = qtyNum / daysNum;
      // Round daily rate to 2 decimals to avoid 0.6666666666 noise.
      // parseFloat round-trip strips trailing zeros: 1.50 → "1.5", 0.67 → "0.67".
      const perDayStr = Number.isInteger(perDay)
        ? String(perDay)
        : Number.parseFloat(perDay.toFixed(2)).toString();
      // P2-b (display precision, 2026-06-08): use NHI's own field labels
      // 給藥總量 / 給藥日數 verbatim instead of the previous synthesized
      // English "N dose(s) over M day(s) (≈ R/day)". Two reasons:
      //   1. Faithfulness — the old "dose(s)" word ASSERTED a dispensing
      //      unit ("dose") that NHI never supplies (IHKE3306S02 ships only
      //      給藥總量 + 給藥日數, no unit field). 給藥總量 is NHI's neutral
      //      count label, so we transport it verbatim and assert no unit.
      //   2. Display — zh-TW patient / clinician context matches the
      //      bundle's patient-facing .text convention (drug_name_zh,
      //      drug_class_zh, reasonCode 繁中).
      // The per-day figure stays a clearly-labelled 平均 (average) derived
      // value; every number here is 100% NHI-sourced (給藥總量 ÷ 給藥日數).
      dosage.text = `給藥總量 ${qtyStr}，給藥日數 ${daysNum} 天（平均每日 ${perDayStr}）`;
    }
  }
  if (raw.route) {
    dosage.route = {
      coding: [{ system: "http://snomed.info/sct", display: raw.route }],
    };
  }
  if (Object.keys(dosage).length > 0) {
    resource.dosageInstruction = [dosage];
  }

  // dispenseRequest with quantity + supply duration when present.
  const dr: Record<string, any> = {};
  const qtyRaw = raw.quantity;
  if (qtyRaw !== null && qtyRaw !== undefined && qtyRaw !== "") {
    const qtyNum = Number.parseFloat(String(qtyRaw).replace(/,/g, ""));
    if (Number.isFinite(qtyNum)) {
      // P2-b (2026-06-08): emit value WITHOUT a unit. NHI IHKE3306S02
      // ships 給藥總量 as a bare count with NO dispensing-unit field
      // (drug obj = drug_name / order_code / order_drug_day / order_qty /
      // act only). The real unit (錠 / mL / 支 / 包 …) depends on the
      // dosage form and cannot be derived reliably from the drug code, so
      // per the faithful-transport rule we do NOT fabricate one — a bare
      // SimpleQuantity.value is FHIR R4 valid. The dosageInstruction text
      // carries the 給藥總量 label so consumers still know the number is
      // a "total dispensed quantity".
      dr.quantity = { value: qtyNum };
    }
  }
  if (raw.duration_days) {
    const days = Number.parseInt(String(raw.duration_days), 10);
    if (Number.isFinite(days)) {
      dr.expectedSupplyDuration = {
        value: days,
        unit: "days",
        system: "http://unitsofmeasure.org",
        code: "d",
      };
    }
  }
  // Inpatient drugs: NHI bundles every drug used during an admission into
  // one row dated to the admission day. authoredOn carries that anchor;
  // validityPeriod expresses the actual usage window [admit, discharge]
  // so SMART apps display "used during stay 5/18-5/22" instead of
  // "all 14 drugs prescribed on 5/18". OPD / 藥局 rows leave end_date
  // empty so this block doesn't fire — single-day prescriptions remain
  // unchanged. The MedicationRequest.dispenseRequest.validityPeriod field
  // is a semantic stretch (its strict definition is the prescription's
  // stale-dating window) but is the closest existing field; we don't
  // emit MedicationAdministration resources.
  const endDate = ((raw.end_date ?? "") as string).trim();
  if (raw.date && endDate && endDate !== raw.date) {
    dr.validityPeriod = {
      start: `${raw.date}T00:00:00+08:00`,
      end: `${endDate}T23:59:59+08:00`,
    };
  }
  if (Object.keys(dr).length > 0) {
    resource.dispenseRequest = dr;
  }

  const indication = ((raw.indication ?? "") as string).trim();
  const indicationZh = ((raw.indication_zh ?? "") as string).trim();
  const indicationCode = ((raw.indication_code ?? "") as string).trim();
  if (indication || indicationZh || indicationCode) {
    const rc: Record<string, any> = {};
    if (indicationCode) {
      rc.coding = [
        {
          system: systems.ICD_10_CM,
          code: normalizeIcd10Cm(indicationCode),
          display: indication || indicationZh || indicationCode,
        },
      ];
    }
    // Patient-facing reasonCode text: prefer 繁中 ICD description, fall
    // back to English, then to just the code. Always prefixed with the
    // code so SMART app rendering keeps "<code> <name>" shape.
    const nameZh = indicationZh || indication;
    if (nameZh) {
      rc.text = indicationCode ? `${indicationCode} ${nameZh}`.trim() : nameZh;
    }
    resource.reasonCode = [rc];
  }

  return resource;
}

/**
 * Group-aware medication mapper that dedupes 中英 雙語 duplicates.
 *
 * Strategy:
 *   1. Compute canonical key per drug name (longest English chunk).
 *   2. Group by (date, canonical_key). Keep ONE entry per group,
 *      preferring the form with FEWER CJK characters (English brand
 *      name — clinicians scan English first).
 *   3. Map each kept entry through mapMedicationRequest.
 *
 * Note: Python comment says "more CJK" but the code uses `<` (fewer);
 * we preserve the actual code behaviour to keep parity.
 */
export function mapMedicationsDedup(rawItems: any[], patientId: string): Record<string, any>[] {
  const byKey = new Map<string, Record<string, any>>();
  for (const item of rawItems) {
    if (!item || typeof item !== "object") continue;
    const drugName = ((item.drug_name ?? "") as string).trim();
    if (!drugName) continue;
    const datePart = ((item.date ?? "") as string).slice(0, 10);
    const key = `${datePart}|${canonicalDrugKey(drugName)}`;
    const existing = byKey.get(key);
    if (existing === undefined) {
      byKey.set(key, item);
    } else {
      // Prefer the form with FEWER CJK characters (English brand name).
      if (cjkChars(drugName) < cjkChars(existing.drug_name ?? "")) {
        byKey.set(key, item);
      }
    }
  }

  const out: Record<string, any>[] = [];
  for (const item of byKey.values()) {
    const m = mapMedicationRequest(item, patientId);
    if (m !== null) out.push(m);
  }
  return out;
}
